src/proto/uprotocol/uauthority.rs

```diff
@@ -11,16 +11,104 @@
  * SPDX-License-Identifier: Apache-2.0
  ********************************************************************************/
 
+use bytes::BufMut;
+
 use crate::uprotocol::uri::uauthority::Number;
 use crate::uprotocol::UAuthority;
 
-pub use crate::uri::validator::ValidationError;
+pub use crate::proto::uprotocol::uuri::SerializationError;
+use crate::uri::validator::ValidationError;
 
 const REMOTE_IPV4_BYTES: usize = 4;
 const REMOTE_IPV6_BYTES: usize = 16;
 const REMOTE_ID_MINIMUM_BYTES: usize = 1;
 const REMOTE_ID_MAXIMUM_BYTES: usize = 255;
 
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum AddressType {
+    Local = 0,
+    IPv4 = 1,
+    IPv6 = 2,
+    ID = 3,
+}
+
+impl AddressType {
+    pub fn value(self) -> u8 {
+        self as u8
+    }
+}
+
+impl TryFrom<u8> for AddressType {
+    type Error = SerializationError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(AddressType::Local),
+            1 => Ok(AddressType::IPv4),
+            2 => Ok(AddressType::IPv6),
+            3 => Ok(AddressType::ID),
+            _ => Err(SerializationError::new(format!(
+                "unknown address type ID [{}]",
+                value
+            ))),
+        }
+    }
+}
+
+impl TryFrom<i32> for AddressType {
+    type Error = SerializationError;
+
+    fn try_from(value: i32) -> Result<Self, Self::Error> {
+        if let Ok(v) = u8::try_from(value) {
+            Self::try_from(v)
+        } else {
+            Err(SerializationError::new(format!(
+                "unknown address type ID [{}]",
+                value
+            )))
+        }
+    }
+}
+
+impl TryFrom<&UAuthority> for AddressType {
+    type Error = SerializationError;
+
+    fn try_from(authority: &UAuthority) -> Result<Self, Self::Error> {
+        if authority.has_id() {
+            Ok(AddressType::ID)
+        } else if authority.has_ip() {
+            match authority.ip().len() {
+                REMOTE_IPV4_BYTES => Ok(AddressType::IPv4),
+                REMOTE_IPV6_BYTES => Ok(AddressType::IPv6),
+                _ => return Err(SerializationError::new("Invalid IP address length")),
+            }
+        } else {
+            Ok(AddressType::Local)
+        }
+    }
+}
+
+impl TryFrom<&UAuthority> for Vec<u8> {
+    type Error = SerializationError;
+
+    fn try_from(authority: &UAuthority) -> Result<Self, Self::Error> {
```

Comment:
FMPOV this needs documentation regarding the format of the byte array being returned, e.g. that this complies with the micro URI serialization format ...

Reply:
Done